### PR TITLE
Remove `:type` lines now sphinx-autoapi supports typehints

### PR DESCRIPTION
**Pull request by ashb**
Since we have no updated sphinx-autoapi to a more recent version it supports showing type hints in the documentation, so we don't need to have the type hints _and_ the `:type` lines -- which is good, as the ones in the doc strings are easy to get out of date!

The following settings have been set:

- `autodoc_typehints = 'description'` -- show types in description (where previous `:type` used to show up)

- `autodoc_typehints_description_target = 'documented'` -- only link to types that are documented. (Without this we have some missing return types that aren't documented, and aren't linked to in our current python API docs, so this caused a build failure)

- `autodoc_typehints_format = 'short'` -- Shorten type hints where possible, i.e. `StringIO` instead of `io.StringIO`

Before (https://airflow.apache.org/docs/apache-airflow/stable/_api/airflow/models/baseoperator/index.html#airflow.models.baseoperator.BaseOperator)

image

After

image

---
**^ Add meaningful description above**

Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/CONTRIBUTING.rst#pull-request-guidelines)** for more information.
In case of fundamental code change, Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvements+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in [UPDATING.md](https://github.com/apache/airflow/blob/main/UPDATING.md).

### CODE REVIEW

**Comment by potiuk** [comment]
HELL YEAH! 

BTW. I guess the main reason you did it is to finally reach negative net number of lines changed in Airflow ?

**Review by potiuk — APPROVED** [review]

**Comment by ashb** [comment]
I mean that was certainly _part_ of the reason. A big part.

It just happens to help that it's also the right thing to do 😁

**Comment by potiuk** [comment]
For some reason building image was cancelled - I reopened to retrigged the build

**Comment by potiuk** [comment]
And removed "full_tests_needed" label too.

**Review by kaxil — APPROVED** [review]

**Comment by potiuk** [comment]
I think there is a "race condition" with just fixed build image failure cancelling which makes the close/reopen workflow cancelling the newly started workflow. I will observe it and see if that's common problem.

**Comment by ashb** [comment]
> I think there is a "race condition" with just fixed build image failure cancelling which makes the close/reopen workflow cancelling the newly started workflow. I will observe it and see if that's common problem.

Yeah, I think I noticed that too today

**Comment by ashb** [comment]
@potiuk Any idea why (total) image re-build would take more than 80 minutes?! https://github.com/apache/airflow/runs/4873253799?check_suite_focus=true

**Comment by ashb** [comment]
Ohhhhh. Pip went in to a huge backtrack loop :( /cc @uranusjr

**Comment by potiuk** [comment]
I think it is induced by the sphinx limitation (main build with eager upgrade does not fail this way). And It's 3.7 only. That should give us a clue.

Likely there is a conflicting dependency in the new sphinx and some other dependency limit. There are some discussions about this behaviour (for example https://github.com/pypa/pip/issues/10655) to get quick faild instead of backtracking. I guess we will need to find out what is the conflict there and fix it - current pip version does not provide a clear reason for the conflict.

**Comment by ashb** [comment]
@potiuk I'm wondering about just using 3.9 to build the docs instead of 3.7? That seems easier than tracking down the dep backtracking issue I think.

WDYT?

**Comment by potiuk** [comment]
> @potiuk I'm wondering about just using 3.9 to build the docs instead of 3.7? That seems easier than tracking down the dep backtracking issue I think.

Unfortunately, there are some consequences of that. This would also mean that we have to switch to 3.9 image to run static checks in the "regular PR case". Currently for those cases we only build 3.7 image and both - documentation building and static checks have to use that one image. This has a number of implications - for example it might well be that some subtle differences from fake or mypy will be missing because they will not be detected for Python 3.9.

**Comment by ashb** [comment]
Ah right okay.

Well I _might_ have found the backtrack anyway 🤞🏻

File: setup.cfg [hunk 1]
```diff
@@ -92,12 +92,6 @@ install_requires =
     cryptography>=0.9.3
     deprecated>=1.2.13
     dill>=0.2.2, <0.4
-    # Sphinx RTD theme 0.5.2. introduced limitation to docutils to account for some docutils markup
-    # change:
-    #      https://github.com/readthedocs/sphinx_rtd_theme/issues/1112
-    # This limitation can be removed after this issue is closed:
-    #      https://github.com/readthedocs/sphinx_rtd_theme/issues/1115
-    docutils<0.17
```
Comment (ashb) [inline setup.cfg]: I don't think this one caused the backtrack, but it's not needed anymore either way.

Comment (potiuk) [reply setup.cfg]: I think the "lazy-object-proxy' in Dockerfile.ci might be the reason  - see #20979

Comment (ashb) [reply setup.cfg]: I really wish this process was more predictable @uranusjr (but I can't think of any obvious way of making it more so)

Comment (potiuk) [reply setup.cfg]: Agree. This ain't easy thing to do. But possibly making eager-update take into account the extras in "one pass" might be possible @uranusjr ? Not sure if that's possible, or if there are some reasons for that. I could possibly open an issue for that, but I am afraid my issues might not be too welcomed as my persistency in improving warning message in the `root` case is seen as a bit `toxic` (even if it is a real issue to tackle).

Comment (ashb) [reply setup.cfg]: Anyway, I from localy testing it appears that the backtrack is removed just by adding the restriction on importlib-metadata, and isn't anything to do with removing this docutils

Comment (potiuk) [reply setup.cfg]: importlib-metadata was my next guess :)

Comment (ashb) [reply setup.cfg]: Either way, 3.7 image is pushing now so found and fixed the backtrack

**Comment by potiuk** [comment]
Woohoo!  seems it's working !

**Review by potiuk — APPROVED** [review]

**Comment by potiuk** [comment]
> Looks like some spelling errors (aka ClassNames) coming in via the types now need to be added to dictionary

But images are ok, all tests pass and all looks good besides :)

**Comment by ashb** [comment]
Right, I think I've got all the type-induced spelling errors, unless one crept back in when rebasing 🤞🏻

**Comment by potiuk** [comment]
2 errors left @ashb !

**Comment by potiuk** [comment]
And conflicts with my cloud formation change :(

**Comment by ashb** [comment]
> And conflicts with my cloud formation change :( 

Fixed the errors.

Has your change merged?

**Comment by ashb** [comment]
@potiuk Any idea what failed with the 'Constraints" job https://github.com/apache/airflow/runs/4888325400?check_suite_focus=true

(I thought we only ran that on main, not on PRs?)

**Comment by ashb** [comment]
Oh there's the error much futher up:

```
#29 8.356 The conflict is caused by:
#29 8.356     apache-airflow[devel-ci] 2.3.0.dev0 depends on sphinx<5.0.0 and >=4.4.0
#29 8.356     The user requested (constraint) sphinx==4.3.2
```

Ummmm, yeah, something is wrong about constraint generation now, as it's trying to generate new constraints while using the old file?!

**Comment by potiuk** [comment]
Rebase ?

**Comment by ashb** [comment]
Already up to date.

**Comment by potiuk** [comment]
Hmm. Will take a look in a bit

**Comment by potiuk** [comment]
We can merge the change now. The problem was because of one line missing in the latest refactor: 
Fix is here: https://github.com/apache/airflow/pull/21000

**Comment by potiuk** [comment]
BTW. This time the round number is mine ;)

**Comment by ashb** [comment]
> We can merge the change now. The problem was because of one line missing in the latest refactor: Fix is here: #21000

Thanks, I had a feeling it would be something like this but didn't want to break main if it was a bigger problem.

**Pull request merged**